micrometer-core/src/main/java/io/micrometer/core/instrument/MeterRegistry.java

```diff
@@ -695,7 +723,8 @@ public Meter remove(Meter meter) {
     @Incubating(since = "1.3.16")
     @Nullable
     public Meter removeByPreFilterId(Meter.Id preFilterId) {
-        return remove(getMappedId(preFilterId));
+        final Meter meterToRemove = preFilterIdToMeterMap.get(preFilterId);
```

Comment:
In case of a meterFilter being added to the registry, before this change this would apply all the latest meter filters and remove only the mapped Id.

Now, we are removing this based on preFilterId, shouldn't this remove both instances, since both would have the same preFilterId? The current implementation would remove only the mappedId and leave the initially mapped Id orphan without a preFilterId reference. This might get tricky as we would have removed the old mapping  when we applied the new mappings.

Reply:
You could also argue that since the mapping changed the old mapping is no longer valid and the old mapping being orphan is expected.

But in this case, I would rather see the orphan meter getting removed rather than sitting idle. But, then there is the problem you shared with references being held and interactions on the orphan meter.

Reply:
I believe the behavior is the same before and after unless I'm missing some case. I wrote this test, for example, and it passes before and after this PR.

```java
@Test
void removeByPreFilterIdAfterAddingFilter() {
    Counter c1 = registry.counter("counter");
    registry.config().commonTags("common", "tag");
    Counter c2 = registry.counter("counter");

    assertThat(registry.removeByPreFilterId(c1.getId())).isSameAs(c2).isNotSameAs(c1);
    assertThat(registry.getMeters()).containsExactly(c1);
}
```

> But, then there is the problem you shared with references being held and interactions on the orphan meter.

Because of this, I don't think we should be changing this behavior yet. I can understand how the behavior can be surprising from a perspective, but changing it has more of a chance of being surprising to someone who (perhaps unknowingly) depends on the current behavior.